### ISSUE TITLE
bash-completion not working properly with directories with spaces in them when using some commands

### ISSUE THREAD

**Issue by jasonp019** [scop/bash-completion#446]
Good evening!

Unfortunately, I need to report an issue with Ununtu 20.04's bash-completion (bash_completion version 2.10). It can't properly expand directories with spaces when there are multiple options when using some commands. Here is an example:

$ mkdir -p "p00p/p00p 1"
$ mkdir -p "p00p/p00p 2"
$ ls -d p00p/p00p\ 1 p00p/p00p\ 2
'p00p/p00p 1' 'p00p/p00p 2'
$ ls p00p/p0<tab><tab>
p00p 1/ p00p 2/
$ ls p00p/p0<tab>1<tab> # <-- works: yields the expanded command: ls p00p/p00p\ 1/
$ wine p00p/p0<tab><tab> # <-- does NOT work: fails to show possible expansions: p00p 1/ p00p 2/
$ wine p00p/p0<tab>1<tab> # <-- does NOT work: fails to expand to: wine p00p/p00p\ 1/

I temporarily fixed it by patching the portion of bash_completion which invokes plusdirs with the version from my raspberry pi OS (bash_completion version 2.8). I've attached the portion of the patch that fixes it.

Unfortunately, to apply the patch with patch -p0, I had to type in the name of the file to patch because the patch I saved was only a subset of the patch produced by "diff -u -w" and my diff-fu isn't good enough to know how to fix it. But, providing the name of the file to be patched when prompted (/usr/share/bash-completion/bash_completion) made it work and after I sourced the new file the issue was fixed.

Thus, this appears to be a regression of sorts but my bash programming skills aren't good enough to fix it myself while ensuring that any functionality provided by the portions that I replaced isn't lost.

Thank you!

Very sincerely,

Jason A. Pfeil

**Comment by jasonp019** [comment]
[patch.txt](https://github.com/scop/bash-completion/files/5211927/patch.txt)

**Comment by scop** [comment]
> $ wine p00p/p01 # <-- does NOT work: fails to expand to: wine p00p/p00p\ 1/

In a scenario containing `p00p/p00p 1` and `p00p/p00p 2` dirs and nothing else in `p00p/`, completing the above should not expand to anything; there's nothing starting with `p01` in `p00p/`. Competing `wine p00p/p0`<kbd>Tab</kbd> there works properly for me (gives  `p00p/p00p 1/ p00p/p00p 2/`), but `wine p00p/p00p\ 1`<kbd>Tab</kbd> does not seem to append the trailing slash. This is with the current master version of bash-completion here, on Ubuntu 18.04.

`_filedir` is a hairy beast. We do have a bunch of test cases for it in the test suite. Alas, your patch does not apply to the current bash_completion master version so it would need to be updated before it can tried out, and the changes are not easily evaluated just by reading the patch so trying out is necessary.

**Comment by zfigura** [comment]
I notice a similar problem with "wine". Given directories "foo bar" and "foo qux", `wine f`Tab will autocomplete to `foo\ `, but pressing Tab again won't display options, and `wine foo\ b`Tab won't autocomplete the directory. In fact, it seems to fail if there's an (escaped) space anywhere, i.e. `wine foo\ bar/f` won't autocomplete to `wine foo\ bar/foo`.

This is especially annoying when trying to autocomplete e.g. `~/.wine/drive_c/Program\ Files`.

**Comment by twotabbies** [comment]
Here is another reproduction case that I hope will be clearer. I have a lot of music arranged as /music/ARTIST/ALBUM. For example
```
ls /music/Mike\ Oldfield/Tubular\ Bells/*.mp3
'/music/Mike Oldfield/Tubular Bells/01 - Part One.mp3'  '/music/Mike Oldfield/Tubular Bells/02 - Part Two.mp3'
```
If I try to use tab completion to play one of these tracks, then tab completion works for the first level of expansion, but not the second. That is, `mplayer /music/Mike<tab>` autocompletes to `mplayer /music/Mike\ Oldfield/`, but then `mplayer /music/Mike\ Oldfield/Tu<tab>` does not autocomplete.

The issue seems to be incorrect handling of the spaces in the first-level directory. Tracing with `set -x` ends up showing (this is just a snippet of the trace that I hope is informative)
```
+ printf -v quoted %q '/music/Mike\ Oldfield/Tu'
+ toks+=($(compgen "${opts[@]}" -- $quoted))
++ compgen -f -X '!*.@(@(m?(j)p?(e)g|M?(J)P?... long regexp trimmed ...' -o plusdirs -- '/music/Mike\\\ Oldfield/Tu'
```
so the already-quoted input is being re-quoted - note the triple backslash in the argument being passed to `compgen`.

Interestingly, `mplayer '/music/Mike Oldfield/Tu'<tab>` does work, and completes to `mplayer /music/Mike\ Oldfield/Tubular\ Bells/`.

If I use `ls` as the command instead of `mplayer`, then autocomplete works perfectly - I guess because it goes through a different code path since it doesn't need to look for files with a particular set of extensions.

**Comment by Izkata** [comment]
Whatever version of bash-completion is in Ubuntu 18.04 didn't have this bug, it was introduced between then and Ubuntu 20.04

(Why I'm only popping in now:  I keep my personal and work computers on the same version so I can use one to debug the other, we only got the go-ahead to upgrade this year due to unrelated issues)

**Comment by kralge** [comment]
I believe I have the same problem, but with `xpdf` as command when using bash-completion v2.16. Autocompletion works until the first directory with space encountered. After that only `.pdf` files offered and no directory at all.

My "workaround" is to use `ls` to find the actual pdf file I am trying to open, then replace the command with `xpdf`... Would be nice to have a real fix...

**Comment by lpbound** [comment]
This bug is annoying, and I think it should be fixed with high priority.

I think this issue is closely related to Debian bug #1061831:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1061831

The Debian reproducer is:

```bash
mkdir "dir blank"
touch "dir blank/a.zip"

# Works:
ls dir\ bla<TAB>

# Fails:
unzip dir\ bla<TAB>
```

I tested the Debian reproducer against upstream 2.17.0 and 2.18.0, and the problem still appears to be present.

I did some additional AI-assisted analysis of this bug. In a nutshell, this is its likely extent and explanation:

> [!NOTE]
> Bash-completion may fail for command-specific filename completions that filter candidates by file type or pattern. This happens when a directory/path-prefix component contains shell-escaped characters such as backslash-escaped spaces.
>
> The likely technical cause is that this prefix is internally escaped a second time, producing strings that no longer match the real filesystem path.

Looking at the current code, the common operation behind the failure seems to be a call to `_comp_quote_compgen` on `cur`, where `cur` may already contain readline/shell escaping. The resulting quoted string is then used as the matching prefix for `_comp_compgen -c`.

In bash-completion 2.18.0, `_comp_quote_compgen` appears to be called in only two relevant filename-completion branches, and both seem affected by this issue:

* [`_comp_compgen_filedir()`](https://github.com/scop/bash-completion/blob/2.18.0/bash_completion#L1344-L1364), relevant to the `plusdirs` branch discussed in this issue:

```bash
local REPLY
_comp_quote_compgen "${cur-}"
local _quoted=$REPLY
...
_comp_compgen -v toks -c "$_quoted" -- "${_opts[@]}"
```

* [`_comp_compgen_filedir_xspec()`](https://github.com/scop/bash-completion/blob/2.18.0/bash_completion#L3281-L3300), relevant to the `xspec` branch used by pattern-filtered completions such as the Debian `unzip` reproducer:

```bash
local REPLY
_comp_quote_compgen "$cur"
local quoted=$REPLY
...
_comp_compgen -v toks -c "$quoted" -- -d
...
_comp_compgen -av toks -c "$quoted" -- -f -X "$xspec"
```

So this may be viewed either as a bug in these two callers, or as a problem in the contract/implementation of `_comp_quote_compgen`: it is being used on readline-escaped `cur` values, but does not first convert them back to literal pathname prefixes before preparing them for `compgen`.

In other words, the problem does not seem to be `plusdirs` alone, nor `xspec` alone. Both manifestations appear to share the same underlying issue: an already shell-escaped path prefix is escaped again before matching, and the resulting over-escaped prefix no longer matches the real filesystem path.

**Comment by akinomyoga** [comment]
The quoting was introduced in 7fcc788d451122d26ff28601cde6cf8a207b8a3f, f98bd24da482d6a1c2df22a30bccc6207fa4c3ec, and 5b304bb4d6e79781e22fada5bb6a056f35984938. This seems to be a port of the changes in Debian bash-completion for the discussion https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=468254, which seems to have attempted to address a similar bug report.

**Comment by lpbound** [comment]
I wonder whether the fix should be to dequote `cur` before it is passed to `_comp_quote_compgen`, e.g. by using `_comp_dequote_incomplete` in the two affected filename-completion branches. Conceptually, the matching prefix should first be converted from readline/shell-escaped form to the literal filesystem prefix, and only then quoted as needed for `compgen`.

Another possibility would be to change the implementation of `_comp_quote_compgen` itself by dequoting its argument at the beginning. But I do not have enough overview of the code to judge whether that would be safe for all callers.

**Comment by lpbound** [comment]
> The quoting was introduced in [7fcc788](https://github.com/scop/bash-completion/commit/7fcc788d451122d26ff28601cde6cf8a207b8a3f), [f98bd24](https://github.com/scop/bash-completion/commit/f98bd24da482d6a1c2df22a30bccc6207fa4c3ec), and [5b304bb](https://github.com/scop/bash-completion/commit/5b304bb4d6e79781e22fada5bb6a056f35984938). This seems to be a port of the changes in Debian bash-completion for the discussion https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=468254, which seems to have attempted to address a similar bug report.

I think these 2008 changes are the origin of the `xspec` branch of this bug. For the `plusdirs` branch – one might argue if it should be considered a separate bug – I found [`da99bc55`](https://github.com/scop/bash-completion/commit/da99bc55954e9f60b9c3a9e9071ff6301d7015cb), `_filedir: avoid duplicate dirs internally, and a compgen -d call for files` from 2019 introducing the problematic `plusdirs` call. This matches the earlier comment that things stopped working at the transition from Ubuntu 18.04 to 20.04.

**Comment by akinomyoga** [comment]
We need to understand the reasoning for the current approach. The suggested fix will fix the present issue you observe, but it might break another specific case. If there is a mistake in the past analysis, we need to identify it and confirm it. To avoid a patchwork, we need to understand the entire picture. It is probably related to the detailed behavior of Bash's internal implementation and the behavior of the `compgen` builtin. **edit**: One needs to care about the difference in unquoting between different options of `compgen`:

```console
$ compgen -f 'dir bla'
dir blank
$ compgen -f 'dir\ bla'
dir blank
$ compgen -d 'dir bla'
dir blank
$ compgen -d 'dir\ bla'
$ compgen -o plusdirs 'dir bla'
dir blank
$ compgen -o plusdirs 'dir\ bla'
$
```

**Comment by lpbound** [comment]
> We need to understand the reasoning for the current approach. The suggested fix will fix the present issue you observe, but it might break another specific case. If there is a mistake in the past analysis, we need to identify it and confirm it. To avoid a patchwork, we need to understand the entire picture. It is probably related to the detailed behavior of Bash's internal implementation and the behavior of the `compgen` builtin. **edit**: One needs to care about the difference in unquoting between different options of `compgen`:
> 
> $ compgen -f 'dir bla'
> dir blank
> $ compgen -f 'dir\ bla'
> dir blank
> $ compgen -d 'dir bla'
> dir blank
> $ compgen -d 'dir\ bla'
> $ compgen -o plusdirs 'dir bla'
> dir blank
> $ compgen -o plusdirs 'dir\ bla'
> $

I agree that a coherent understanding of the past design decisions and intentions would be very valuable. In particular, it would help to clarify the intended contracts of the core helper functions: what form of input they expect, and what form of output they promise, especially for quoted, escaped, half-quoted, and trailing-backslash cases.

At the same time, I wonder whether a fully specified “entire picture” exists. For example, the [`compgen` documentation](https://www.gnu.org/software/bash/manual/html_node/Programmable-Completion-Builtins.html) does not seem to specify in detail what should happen for an escaped prefix such as `dir\ bla`, and your examples show that the current behavior differs between `-f`, `-d`, and `-o plusdirs`. So perhaps a pragmatic way forward would be to encode the expected behavior as unit tests. The relevant test matrix could include at least:

* escaped vs quoted prefixes, e.g. `dir\ bla`, `'dir bla'`, `"dir bla"`;
* incomplete quoting / trailing backslash cases;
* escaped characters in the final filename component vs in an earlier directory component;
* the `_comp_compgen_filedir()` / `plusdirs` branch;
* the `_comp_compgen_filedir_xspec()` / `xspec` branch.

Your example calls of `compgen` show that escaped prefixes are not handled consistently by the different `compgen` modes: `-f` accepts `dir\ bla`, but `-d` and `-o plusdirs` do not. However, all three modes appear to work consistently with the literal filesystem prefix `dir bla`. This supports the idea that the current word should first be converted from readline/shell-escaped form to the literal filesystem prefix before it is used as the matching prefix for these filename-completion branches.

A personal note: I want to emphasize that I consider this a serious bug that deserves high-priority attention. I run into it almost daily, as it affects one of the central functionalities of bash-completion. I strongly hope we can turn the analysis above into a timely fix.

**Comment by lpbound** [comment]
I did some more research on this bug.

(1)
In my opinion, the core reason is that the treatment of escaped characters of `compgen` in directory/plusdirs mode (i.e., `-d` or `-o plusdirs`) is not matching our needs, see the above experiments of akinomyoga.
Accepting this arguably inconsistent behavior of `compgen` as given, the robust fix inside `bash-completion` seems to be to convert the current readline/shell-escaped word into the literal filesystem prefix before passing it to `compgen`.

(2)
As a second component of the bug, there is still the over-quotation of the `compgen` argument, which I had already pointed out in an earlier post. I did some experiments, and I think the extra quotation level done by the function `_comp_quote_compgen` (originally done by `quote_readline` until 2009, then via `_quote_readline_by_ref` until 2023) works as a no-op in the common simple cases, and results in an over-quotation of the already problematic cases; however, those cases would not have worked even without the surplus quotes, for reason (1) above.

For the `_filedir` branch, the user-visible manifestation of the bug exists since 2019, where the machinery started relying on the `compgen -o plusdirs` call. For the `xspec` branch, the problem appears to go back even before the discussed 2008 changes, possibly to the very beginning. In daily use, I guess the `_filedir` branch is the more frequent one, explaining that complaints and bug reports only started when people noticed the worsening since 2019.

**Comment by lpbound** [comment]
For a fix, we can replace the problematic and essentially unnecessary `_comp_quote_compgen` call by `_comp_dequote_incomplete`, which transforms the current readline/shell-escaped word into a literal filesystem prefix, while also handling incomplete quotes and incomplete escapes (unmatched trailing backslashes)

Hence, in `bash_completion`, the lines in `_comp_compgen_filedir()`:

```bash
local REPLY
_comp_quote_compgen "${cur-}"
local _quoted=$REPLY
_comp_unlocal REPLY

# work around bash-4.2 where compgen -f "''" produces nothing.
[[ $_quoted == "''" ]] && _quoted=""
```

should be replaced by:

```bash
local REPLY
_comp_dequote_incomplete "${cur-}" || true
local _quoted=${REPLY[0]-}
_comp_unlocal REPLY
```

And the lines in `_comp_compgen_filedir_xspec()`:

```bash
local REPLY
_comp_quote_compgen "$cur"
local quoted=$REPLY
```

should be replaced by:

```bash
local REPLY
_comp_dequote_incomplete "$cur" || true
local quoted=${REPLY[0]-}
```

The important technical detail is `${REPLY[0]-}`: `_comp_dequote_incomplete` returns its result via array-style `REPLY`, unlike `_comp_quote_compgen`, which returns a scalar `REPLY`.

Alternatives would be to introduce a small helper function that centralizes the `_comp_dequote_incomplete` call, or to make `_comp_compgen` itself responsible for normalizing readline/shell-escaped filename prefixes before calling `compgen`.

I did extended AI-assisted testing of the suggested changes on bash-completion 2.18.0, covering the cases in the test matrix I suggested above, including the current GitHub reproducers and the historical Debian #468254-style cases. The tested cases all behaved as expected.

Would it be okay if I prepared a pull request with these changes?

**Comment by lpbound** [comment]
Things are getting really spicy now!

First of all, my statement "The tested cases all behaved as expected." was a bit premature. That test directly exercised the internal bash-completion helper functions, but not the full interactive programmable-completion path. Repeating the test in a real-world interactive bash session changes the picture. The patch as suggested above fixes the main escaped-prefix problem, but unfortunately, the following 9 test cases fail (those were derived from the test matrix I gave in an earlier post):

- `wine dir\'quo<TAB>` with the existing file `dir'quote/foo.exe`
- `unzip dir\'quo<TAB>` with the existing file `dir'quote/a.zip`
- `wine dir\"quo<TAB>` with the existing file `dir"quote/foo.exe`
- `unzip dir\"quo<TAB>` with the existing file `dir"quote/a.zip`
- `wine dir\\ba<TAB>` with the existing file `dir\back/foo.exe`
- `unzip dir\\ba<TAB>` with the existing file `dir\back/a.zip`
- `unzip a\'quo<TAB>` with the existing file `a'quote.zip`
- `unzip a\"quo<TAB>` with the existing file `a"quote.zip`
- `unzip a\\ba<TAB>` with the existing file `a\back.zip`

The reason seems to be that in these cases, `compgen` expects its word argument in quoted/escaped form. Hence, completion of these cases, which were passed in literal form to `compgen`, fails.

Adding the `_comp_quote_compgen` call again to the `bash-completion` code, while keeping `_comp_dequote_incomplete` nested inside it, makes the above 9 cases work correctly. However, this unconditional dequote-then-quote variant creates 5 other failures, namely:

- `wine 'dir bla<TAB>` with the existing file `dir blank/foo.exe`
- `unzip 'dir bla<TAB>` with the existing file `dir blank/a.zip`
- `wine "dir bla<TAB>` with the existing file `dir blank/foo.exe`
- `unzip "dir bla<TAB>` with the existing file `dir blank/a.zip`
- `unzip "<TAB>` with the existing file `a.zip`

For these cases, it turns out that `compgen` expected the input string in literal form, resulting in faulty behaviour given the quoted input.

So the actual situation seems to be: in a real programmable-completion context, `compgen` sometimes expects the word in readline/shell-escaped form and sometimes in literal/dequoted form. What the heck is going on?

For a perfect patch, we need to know when to quote and when not to quote. As a side note: the first 9 cases already fail with the current stock version, so while the pure `_comp_dequote_incomplete` patch suggested in my previous post does not fix everything, it would still be a substantial improvement over the current situation.

An investigation of the `compgen` code in bash gives the following key section in `pcomplete.c`, in `pcomp_filename_completion_function()`. It was introduced by the official upstream [Bash patch `bash43-010`](https://ftp.gnu.org/gnu/bash/bash-4.3-patches/bash43-010) in 2014.

```c
/* Another mismatched assumption by bash-completion.  If compgen is being
   run as part of bash-completion, and the argument to compgen is not
   the same as the word originally passed to the programmable completion
   code, dequote the argument if it has quote characters.  It's an
   attempt to detect when bash-completion is quoting its filename
   argument before calling compgen. */

/* We could check whether gen_shell_function_matches is in the call
   stack by checking whether the gen-shell-function-matches tag is in
   the unwind-protect stack, but there's no function to do that yet.
   We could simply check whether we're executing in a function by
   checking variable_context, and may end up doing that. */
else if (iscompgen && iscompleting && rl_filename_dequoting_function &&
         pcomp_curtxt && text &&
         STREQ (pcomp_curtxt, text) == 0 &&
         variable_context &&
         sh_contains_quotes (text)) /* guess */
  dfn = (*rl_filename_dequoting_function) ((char *)text,
                                           rl_completion_quote_character);
```

I hope you didn't miss the comment! It is pretty revealing: at some point, certain problematic assumptions in `bash-completion` were detected, but apparently, it was considered more practical to compensate for that inside `bash` itself than to wait for the underlying issue to be fixed in `bash-completion`. From a pragmatic point of view, that may have been the best approach at the time, but it also means that `bash-completion` is now interacting with a `bash`-side heuristic explicitly introduced to work around `bash-completion`'s own quoting assumptions. This should be another motivation to finally fix the issue on the `bash-completion` side.

In detail, `bash` now contains a special heuristic for the exact situation where `bash-completion` quotes the filename argument before calling `compgen`. The condition is triggered when `compgen` runs during active programmable completion, the word passed to `compgen` differs from the original word being completed, and that word contains one of `'`, `"`, or `\`.

For a fix, my suggestion is:
1. I think the immediate user-visible bug should be fixed now, not deferred to a larger cleanup or redesign. This issue has been known for years, and it affects ordinary interactive filename completion in a serious way. I suggest this strategy: After applying `_comp_dequote_incomplete`, anticipate the above condition to decide whether `compgen` expects the argument in quoted or in literal form, and prepare it accordingly. This should probably be done within the `_comp_compgen` wrapper function.
2. In the longer term, there should be some coordination with the `bash` package, to get both sides in a clean state and to remove code that reacts to some inconsistent/buggy behavior on the other side.

**Comment by lpbound** [comment]
Based on the ideas in my last comment, I have now created a [patch](https://github.com/user-attachments/files/30525133/compgen-quote-dequote.patch). It applies to a clean 2.18.0 checkout.

Let me state clearly that I created it in an AI-assisted way. I would not have been able to write it entirely on my own, as I know only the basics of bash scripting and, frankly, I am always a bit scared by it.

However, I certainly did not blindly accept the resulting code. It went through several rounds of reading, questioning, testing, and simplification of the logic.

I also ran an AI-assisted test of the patch. It was done in a real interactive `bash` session, not only at helper-function level. The patch was tested against 36 test cases, covering:

- the current `wine` / `mplayer` escaped-path reproducers,
- `unzip` / xspec-style cases related to zip-file completion,
- historical whitespace-style cases from the 2008 Debian bug discussion,
- border cases arising from the test matrix I suggested in an earlier post, including the 9 + 5 problematic cases from my last comment,
- empty and near-empty prefixes,
- and two additional open-quote edge cases.

The result was:

```text
stock 2.18.0: pass=20 fail=16 total=36
patched:      pass=36 fail=0  total=36
```

What do you think about this patch?
Would this be an acceptable direction for a pull request?

**Comment by lpbound** [comment]
I found the test-suite only by now, shame on me. And it shows that my latest patch proposal is not perfect either, leading to a few regressions: certain cases with open quotes are not handled correctly. Sigh.

**Comment by lpbound** [comment]
My latest conclusion is that, under the given restrictions on `compgen` in `-d` or `-o plusdirs` mode, there might not exist a static patch that handles all possible mixed quoting states solely by choosing the right `_cur` string.

Exercise: Assume we have the literal pathname

```text
a"bcd/e.zip
```

and the user types:
```text
unzip a\"b"c<TAB>
```

Question: Find a suitable string `_cur` for which `compgen` produces one of the the expected completions
```text
unzip a\"b"cd"/
```
or
```text
unzip a\"b"cd/"
```
or (quote being still open)
```text
unzip a\"b"cd/
```

The solution might be: there is no such `_cur`.

Rationale: To avoid the current `compgen` restrictions, we cannot pass the first `"` in escaped form as `\"`, because this is exactly the kind of escaped character on which `compgen -d` / `compgen -o plusdirs` currently fails. So this first `"` has to appear as an unescaped literal filename character. But then we no longer have a way to distinguish the second `"` symbol from the first one, although it has a different role, namely being the opening incomplete quote mark. In other words, the two syntactically different `"` characters collapse into the same character in the string passed to `compgen`.

**Comment by lpbound** [comment]
To get further insight, I did an AI-assisted research of the historical development of the relevant parts of the `compgen` code in the [bash](https://www.gnu.org/software/bash/) package.
The points I consider most relevant are:

- Around **2000**, `compgen -f` and `compgen -d` already used separate internal code paths. `CA_FILE` goes through `pcomp_filename_completion_function`, while `CA_DIRECTORY` goes through `bash_directory_completion_matches`. So file completion and directory completion were structurally separate and probably carried differing semantics from the beginning of programmable completion.

- Around **2004**, `plusdirs` was added (https://cgit.git.savannah.gnu.org/cgit/bash.git/commit/?id=b80f6443b6b7b620c7272664c66ecb0b120a0998), and it reuses the directory-completion path by synthesizing a `CA_DIRECTORY` completion. So the directory part of `compgen -f -o plusdirs` is not just ordinary `-f` behavior; it goes through the `-d`-style machinery.

- Before **2014** / Bash 4.3, bash-completion appears to have relied on Bash/Readline dequoting quoted arguments passed to `compgen`. I would not call this a clean documented contract, but it clearly was an assumption bash-completion made.

- In **2014**, for Bash 4.3, this behavior was changed: the official upstream [Bash patch `bash43-010`](https://ftp.gnu.org/gnu/bash/bash-4.3-patches/bash43-010) says that Bash 4.3 “changed to dequote the argument to programmable completion only under certain circumstances,” with the goal of making `compgen` more consistent between command-line use and use inside a completion function. This Bash 4.3 change broke assumptions that bash-completion relied on, and I suspect that this change is one important origin of today’s trouble.

- To mitigate that effect, the same patch already added two explicit compatibility workarounds for bash-completion: one for quoted empty strings like `''`, and one for quoted filename arguments passed to `compgen`. The source comments are explicit: “Intended to solve a mismatched assumption by bash-completion” and “Another mismatched assumption by bash-completion.” However, I doubt that the underlying heuristic when to trigger that exception perfectly matches bash-completion's logic, given all those problematic border cases we know about today.

- Shortly afterwards, still in **2014**, [Bash patch `bash43-015`](https://ftp.gnu.org/gnu/bash/bash-4.3-patches/bash43-015) fixed a bug in the directory-completion path, `bash_directory_completion_matches`, to avoid double dequoting when `rl_completion_found_quote` is set. This matters because `compgen -d` and the directory part of `-o plusdirs` go through that path, not through the same path as `compgen -f`; before this patch, bash-completion’s over-quoted directory prefixes may therefore have been accidentally tolerated in some cases.

So my current interpretation is: 
From a higher-level perspective, the problem behind bug #442 cannot be reduced to a simple quoting mistake in bash-completion. Rather, it sits at the boundary between bash-completion’s prefix preparation and Bash’s `compgen` / Readline dequoting machinery. The matter is complicated by the fact that this machinery has grown historically: its current rules for when dequoting is triggered are neither simple nor conceptually coherent, and they are not officially documented.
